### PR TITLE
Adding docstrings to all public modules

### PR DESCRIPTION
**Pull request by kenodegard**
### Description

Adds missing module docstrings. These initial docstrings are not particularly detailed but seek to provide a rough sense of what the module is for or how it relates to other modules.

Xref https://github.com/conda/conda/issues/12613
Resolves https://github.com/conda/conda/issues/12668
Depends on #12798 

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [ ] ~Add / update necessary tests?~
- [x] Add / update outdated documentation?

### CODE REVIEW

File: conda/cli/main.py [hunk 1]
```diff
@@ -1,36 +1,6 @@
...
-
-    Packaging
-    =========
-
-    package    : create a conda package in an environment
-
-Additional help for each command can be accessed by using:
-
-    conda <command> -h
-"""
```
Comment (travishathaway) [inline conda/cli/main.py:33]: Wow, this one was ancient! 👏  Weird how many times I've seen this file only to skip this header 😅 .

File: conda/models/channel.py [hunk 2]
```diff
@@ -1,5 +1,15 @@
 # Copyright (C) 2012 Anaconda, Inc
 # SPDX-License-Identifier: BSD-3-Clause
+"""Defines Channel and MultiChannel objects and other channel-related functions.
+
+Object inheritance:
+ ┌─────────┐
```
Comment (travishathaway) [inline conda/models/channel.py:6]: Where's the meta-class? 😏 Just kidding. Looks good 👌.

**Review by travishathaway — APPROVED** [review]
LGTM. I really like the inheritance diagrams 👌.

File: conda/common/decorators.py [hunk 3]
```diff
@@ -1,9 +1,15 @@
 # Copyright (C) 2012 Anaconda, Inc
 # SPDX-License-Identifier: BSD-3-Clause
+"""Common deprecators."""
```
Comment (jaimergp) [inline conda/common/decorators.py:3]:
```suggestion
"""Common decorators."""
```
Did you mean this?

File: conda/resolve.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 # Copyright (C) 2012 Anaconda, Inc
 # SPDX-License-Identifier: BSD-3-Clause
+"""The low-level SAT solver wrapper/interface (conda.core.solver.Solver is a higher-level API)."""
```
Comment (jaimergp) [inline conda/resolve.py:3]: This is only part of the `classic` solver, if that needs to be added.

File: tests/core/test_subdir_data.py [hunk 5]
```diff
@@ -137,70 +136,6 @@ def test_get_index_no_platform_with_offline_cache(self, platform=OVERRIDE_PLATFO
 
 
 class StaticFunctionTests(TestCase):
-    def test_read_mod_and_etag_mod_only(self):
```
Comment (jaimergp) [inline tests/core/test_subdir_data.py:140]: Why is this deleted as part of this PR? 🤔

File: conda/common/README.md
Comment (jaimergp) [inline conda/common/README.md:1]: Should this content be added to `conda.common.__init__` instead?

File: conda/core/README.md
Comment (jaimergp) [inline conda/core/README.md:1]: Should this content (or a variant of it) be added to `conda.core.__init__` instead?

File: conda/gateways/README.md
Comment (jaimergp) [inline conda/gateways/README.md:1]: Should this content be added to `conda.gateways.__init__` instead?

File: conda/models/README.md
Comment (jaimergp) [inline conda/models/README.md:1]: Should this content be added to `conda.models.__init__` instead?

File: conda/models/records.py [hunk 6]
```diff
@@ -1,14 +1,24 @@
...
-+--------------------+                       +--------------+
-
-
+"""Implements the data model for conda packages.
+
+A PackageRecord is the record of a package present in a channel. A PackageCache is the record of a
+downloaded and cached package. A PrefixRecord is the record of a package installed into a conda
+environment.
+
+Object inheritance:
```
Comment (jezdez) [inline conda/models/records.py:9]: Same, let's drop this

File: conda/models/version.py [hunk 7]
```diff
@@ -1,5 +1,18 @@
...
+Object inheritance:
+ ┌──────────┐
+ │ BaseSpec │
+ └┬─────────┘
+  │ ┌─────────────┐
+  ├─┤ VersionSpec │
+  │ └─────────────┘
+  │ ┌──────────────────┐
+  └─┤ BuildNumberMatch │
+    └──────────────────┘
```
Comment (jezdez) [inline conda/models/version.py:14]: same

File: conda/resolve.py (hunk 4 above)
Comment (jezdez) [reply conda/resolve.py:3]: Good idea

File: conda/trust/constants.py [hunk 8]
```diff
@@ -1,15 +1,18 @@
...
-# bootstrapping, the first/backup version of the root of trust metadata.
-# Still, the signatures here are useful for diagnostic purposes, and, more
-# important, to allow self-consistency checks: that helps us avoid breaking the
-# chain of trust if someone accidentally lists the wrong keys down the line. (:
-# The discrepancy can be detected when loading the root data, and we can
-# decline to cache incorrect trust metadata that would make further root
-# updates impossible.
+"""Context trust constants.
+
+You could argue that the signatures being here is not necessary; indeed, we
```
Comment (jezdez) [inline conda/trust/constants.py:5]: @awwad is this still accurate?

**Review by jezdez — CHANGES_REQUESTED** [review]

File: conda/models/prefix_graph.py [hunk 9]
```diff
@@ -1,5 +1,15 @@
 # Copyright (C) 2012 Anaconda, Inc
 # SPDX-License-Identifier: BSD-3-Clause
+"""Implements directed graphs to sort and manipulate packages within a prefix.
+
+Object inheritance:
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply conda/models/prefix_graph.py:5]: Can these diagrams be fenced (triple backticks?). Alternatively, it could use [Mermaid](https://github.com/mgaitan/sphinxcontrib-mermaid) syntax.

File: conda/models/channel.py (hunk 2 above)
Comment (kenodegard) [reply conda/models/channel.py:6]: ah i forgot about that! do we think they're valuable at all? if so i'll move them into a comment below the docstring

cc @jaimergp

File: tests/core/test_subdir_data.py (hunk 5 above)
Comment (kenodegard) [reply tests/core/test_subdir_data.py:140]: minor deprecation that i forgot to pull out into a separate PR, let me do that

File: conda/models/prefix_graph.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jezdez) [reply conda/models/prefix_graph.py:5]: I think you can add mermaid blocks using one of the sphinx extensions, yeah.

File: conda/models/channel.py (hunk 2 above)
Comment (jezdez) [reply conda/models/channel.py:6]: see Jaime's comment below, suggesting to use mermaid for this, I know there are a few Sphinx extensions that can render `.. mermaid::` blocks and there is even a `autoclasstree` directive in https://github.com/mgaitan/sphinxcontrib-mermaid! Demo: https://sphinxcontrib-mermaid-demo.readthedocs.io/

File: conda/common/README.md
Comment (kenodegard) [reply conda/common/README.md:1]: already is from before

File: conda/core/README.md
Comment (kenodegard) [reply conda/core/README.md:1]: already is from before

File: conda/gateways/README.md
Comment (kenodegard) [reply conda/gateways/README.md:1]: already is from before

File: conda/base/README.md [hunk 10]
```diff
@@ -1,18 +0,0 @@
...
-- `conda._vendor`
-- `conda.base`
-- `conda.common`
-
-Modules prohibited from importing `conda.base` are:
-
-- `conda._vendor`
-- `conda.common`
-
-All other `conda` modules may import from `conda.base`.
```
[the remark this answers is not in the document]
Comment (kenodegard) [reply conda/base/README.md:18]: already is from before

File: conda/cli/python_api.py [hunk 11]
```diff
@@ -1,16 +1,19 @@
...
 from ..base.context import context
 from ..common.compat import encode_arguments
 from ..common.io import CaptureTarget, argv, captured
+from ..deprecations import deprecated
 from ..exceptions import conda_exception_handler
 from ..gateways.logging import initialize_std_loggers
 from .conda_argparse import do_call
 from .main import generate_parser
 
+deprecated.module("24.3", "24.9", addendum="Use `conda.testing.conda_cli` instead.")
```
[the remark this answers is not in the document]
Comment (kenodegard) [reply conda/cli/python_api.py:16]: I'm torn over this since the python_api isn't for testing only, let's me move this into a separate PR so we can have a cleaner discussion

File: conda/models/version.py (hunk 7 above)
Comment (kenodegard) [reply conda/models/version.py]: switched to `sphinxcontrib-mermaid`

File: conda/models/records.py (hunk 6 above)
Comment (kenodegard) [reply conda/models/records.py:9]: switched to `sphinxcontrib-mermaid`

File: conda/models/prefix_graph.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kenodegard) [reply conda/models/prefix_graph.py:5]: switched to `sphinxcontrib-mermaid`

File: conda/models/channel.py (hunk 2 above)
Comment (kenodegard) [reply conda/models/channel.py]: switched to `sphinxcontrib-mermaid`

File: conda/cli/main_pip.py [hunk 12]
```diff
@@ -1,5 +1,6 @@
 # Copyright (C) 2012 Anaconda, Inc
 # SPDX-License-Identifier: BSD-3-Clause
+"""Entry point if `conda` has not been properly initialized in the user's shell."""
```
[the remark this answers is not in the document]
Comment (jezdez) [reply conda/cli/main_pip.py:3]:
```suggestion
"""PEP 621 compatible entry point used when `conda init` has not updated the user shell profile."""
```

File: conda/cli/python_api.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jezdez) [reply conda/cli/python_api.py]: Sounds good

File: conda/models/channel.py (hunk 2 above)
Comment (jaimergp) [reply conda/models/channel.py]: Nice!

I think we still need to add the requirement to https://github.com/conda/conda/blob/main/docs/requirements.txt and the package name to https://github.com/conda/conda/blob/main/docs/source/conf.py#L41

@kenodegard

Comment (kenodegard) [reply conda/models/channel.py]: agreed, I'm also experimenting with adding sphinx-autoapi in #12798 ATM

**Review by jezdez — APPROVED** [review]

**Comment by jaimergp** [comment]
Btw, `interrogate` tests would be useful for this kind of work:
https://github.com/econchick/interrogate

> interrogate checks your code base for missing docstrings

**Comment by kenodegard** [comment]
@jaimergp we use `flake8-docstrings`, it'd enabled with pre-commit but we're currently ignoring many of the checks until we have time to address the missing docstrings (like here)

how does `interrogate` compare with `flake8-docstrings`?

**Comment by jaimergp** [comment]
Oh if we have that flake8 plugin then it's ok. interrogate has prettier reports of missing docstrings and a coverage score, I think, but that's it.
